catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/suggestion/MgrsCoordinateProcessor.java

```diff
@@ -0,0 +1,354 @@
+
+  private static final Pattern PATTERN_MGRS_100KM = Pattern.compile("[A-HJ-NP-Z][A-HJ-NP-V]");
+
+  private static final Pattern PATTERN_MGRS_NUMERIC =
+      Pattern.compile("((\\d){1,5}(\\h)+(\\d){1,5}(\\h)+)|((\\d){2,10})");
+
+  private static final Pattern PATTERN_MGRS_ANYPART =
+      Pattern.compile(
+          format(
+              "(%s)|(%s)|(%s)",
+              PATTERN_MGRS_ZONE.pattern(),
+              PATTERN_MGRS_100KM.pattern(),
+              PATTERN_MGRS_NUMERIC.pattern()),
+          Pattern.CASE_INSENSITIVE);
+
+  // usng4j helper constant for backfilling zeros
+  private static final Integer MAX_EASTING_LENGTH = 5;
+
+  // usng4j helper constant for backfilling zeros
+  private static final Integer MAX_NORTHING_LENGTH = 5;
+
+  // This key tells the UI that the geo is on the suggestion itself
+  private static final String LITERAL_SUGGESTION_ID = "LITERAL2";
+
+  private static final String ZERO = "0";
+
+  private static final String WHITESPACE = " ";
+
+  /**
+   * An {@link MgrsPartType} identifies one of the three different component strings that are glued
+   * together to form a valid MGRS coordinate string, including a category {@link #NONE} to be used
+   * when the string is unsuitable for any part of an MGRS coordinate or the type has not been
+   * provided yet.
+   *
+   * <p>Each type corresponds to a regex defined above.
+   */
+  private enum MgrsPartType {
+    NONE,
+    ZONE,
+    HUNDRED_KM,
+    NUMERIC
+  }
+
+  private final CoordinateSystemTranslator translator = new CoordinateSystemTranslatorImpl();
+
+  /**
+   * Given a list of {@link Suggestion}s and the query that yielded them, enhance the list with
+   * additional suggestions based upon the presence of coordinate literals in the query string.
+   *
+   * @param results the list of results to enhance.
+   * @param query the query string to search for coordinate literals.
+   */
+  public void enhanceResults(List<Suggestion> results, String query) {
+    LiteralSuggestion literal = getMgrsSuggestion(query);
+    if (literal != null && literal.hasGeo()) {
+      LOGGER.debug("Adding the MGRS suggestion to results");
+      results.add(0, literal);
+    }
+  }
+
+  /**
+   * Build the MGRS suggestion.
+   *
+   * @param query the query string to search for coordinate literals.
+   * @return a suggestion object with geometry attached to it; a {@link BoundingBox} if only one
+   *     coordinate was detected, or a collection of {@link LatLon} center points otherwise. If no
+   *     MGRS was detected in the query, {@code null} is returned.
+   */
+  @Nullable
+  private LiteralSuggestion getMgrsSuggestion(String query) {
+    final List<String> matches = getMgrsCoordinateStrings(query);
+    if (matches.isEmpty()) {
+      LOGGER.debug("No valid MGRS strings were found or could be inferred from query [{}]", query);
+      return null;
+    }
+    final List<UsngCoordinate> uniqueMatches =
```

Comment:
Can we not use a for loop and prevent having to iteration on `matches` twice?

Reply:
I only see `matches` being iterated through once (on the line below this one); where is the other iteration?

Reply:
Not explicitly, but we are iterating over the collection twice... first we parse the collection into MGRS strings in one stream, and then add `[]` to them in the next stream and collect it into a string.   Not sure why we are going out of our way to iterate twice over a collection if it can be done in one iteration.  This is my question.

Reply:
There are cases where, despite the multiple iteration, the actual computation difference is negligible and we get a more functional result. I'd argue it's a good trade off. But I don't mind changing it.